<a href="https://colab.research.google.com/github/PatrickRuan/3rd-ML100Days/blob/master/homework/Day_066_HW.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from keras.utils import multi_gpu_model
from keras.models import Model
from keras.layers import Input, Dense


a = Input(shape=(32,))
b = Dense(32)(a)
model = Model(inputs=a, outputs=b)

config = model.get_config()
print(config)

Using TensorFlow backend.





{'name': 'model_1', 'layers': [{'name': 'input_1', 'class_name': 'InputLayer', 'config': {'batch_input_shape': (None, 32), 'dtype': 'float32', 'sparse': False, 'name': 'input_1'}, 'inbound_nodes': []}, {'name': 'dense_1', 'class_name': 'Dense', 'config': {'name': 'dense_1', 'trainable': True, 'dtype': 'float32', 'units': 32, 'activation': 'linear', 'use_bias': True, 'kernel_initializer': {'class_name': 'VarianceScaling', 'config': {'scale': 1.0, 'mode': 'fan_avg', 'distribution': 'uniform', 'seed': None}}, 'bias_initializer': {'class_name': 'Zeros', 'config': {}}, 'kernel_regularizer': None, 'bias_regularizer': None, 'activity_regularizer': None, 'kernel_constraint': None, 'bias_constraint': None}, 'inbound_nodes': [[['input_1', 0, 0, {}]]]}], 'input_layers': [['input_1', 0, 0]], 'output_layers': [['dense_1', 0, 0]]}


In [2]:
model.summary()

Model: "model_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         (None, 32)                0         
_________________________________________________________________
dense_1 (Dense)              (None, 32)                1056      
Total params: 1,056
Trainable params: 1,056
Non-trainable params: 0
_________________________________________________________________


# 作業:
    檢查 backend
    檢查 fuzz factor
    設定 Keras 浮點運算為float16

In [4]:
import keras
from keras import backend as K

#檢查 backend
keras.backend.backend()


'tensorflow'

# fuzz factor

數值表達式中使用的模糊因子。
此功能是一組Keras後端功能的一部分，可以對後端張量引擎（例如TensorFlow，CNTK，Theano等）的核心操作進行更低級別的訪問。

In [5]:
#檢查 fuzz factor
keras.backend.epsilon()


1e-07

In [6]:
#設定 Keras 浮點運算為float16
K.set_floatx('float16')
K.floatx()


'float16'

In [7]:
print(keras.__version__)

2.2.5


In [8]:
#  GPU加速测试, True - Windows用戶得到True也沒有關係，因為Anaconda中已經內置了MKL加速庫
import numpy 
id(numpy.dot) == id(numpy.core.multiarray.dot) 

True

# **常見錯誤處理

**常見錯誤：**FutureWarning: Conversion of the second argument of issubdtype from floatto np.floatingis deprecated. In future, it will be treated as np.float64 == np.dtype(float).type.
from ._conv import register_converters as _register_converters
**解決方案：**pip install h5py==2.8 .0rc1，安裝h5py，用於模型的保存和載入

**切換後端**Using TensorFlow backend.
但是keras的backend同時支持tensorflow和theano.
並且默認是tensorflow,

**常見錯誤：**TypeError: softmax() got an unexpected keyword argument 'axis'
**解決方案：**pKeras與tensorflow版本不相符，盡量更新最新版本：pip install keras==2.2

